### PR TITLE
ddl: check unique key constraint for 'alter table add index' on partitioned table

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/9100

### What is changed and how it works?

> Every unique key on the table must use every column in the table's partitioning expression.

We have this check when creating a table, 'alter table add index' should also check this.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @ciscoxll @zimulala @winkyao

File: ddl/db_test.go [hunk 1]
```diff
@@ -2362,3 +2362,11 @@ func (s *testDBSuite) TestAddIndexForGeneratedColumn(c *C) {
 	s.mustExec(c, "alter table t add index idx_y(y1)")
 	s.mustExec(c, "alter table t drop index idx_y")
 }
+
+func (s *testDBSuite) TestIssue9100(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test_db")
+	tk.MustExec("create table employ (a int, b int) partition by range (b) (partition p0 values less than (1));")
+	_, err := tk.Exec("alter table employ add unique index  p_a (a);")
+	c.Assert(err.Error(), Equals, "[ddl:1503]A PRIMARY KEY must include all columns in the table's partitioning function")
```
Comment (jackysp) [inline ddl/db_test.go]: The error message should be `A UNIQUE INDEX must include all columns in the table's partitioning function`?

Comment (zimulala) [reply ddl/db_test.go]: And it seems that the error of `create table employ (a int, b int, unique index(a)) partition by range (b) (partition p0 values less than (1));` also has this problem.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -2509,6 +2509,30 @@ func (d *ddl) CreateIndex(ctx sessionctx.Context, ti ast.Ident, unique bool, ind
...
+		for _, col := range cols {
+			partkeys = append(partkeys, col.ColName.L)
+		}
+		constraints := make(map[string]struct{})
+		for _, uniqCol := range idxColNames {
+			constraints[uniqCol.Column.Name.L] = struct{}{}
+		}
+		if !checkConstraintIncludePartKey(partkeys, constraints) {
+			return ErrUniqueKeyNeedAllFieldsInPf.GenWithStackByArgs(primarykey)
+		}
```
Comment (zimulala) [inline ddl/ddl_api.go]: Could this be extracted into a function to partition.go?

File: ddl/db_test.go (hunk 1 above)
Comment (ciscoxll) [reply ddl/db_test.go]: Can you add this tests ?
```sql
mysql> CREATE TABLE t1 (
    ->     col1 INT NOT NULL,
    ->     col2 DATE NOT NULL,
    ->     col3 INT NOT NULL,
    ->     col4 INT NOT NULL,
    ->     UNIQUE KEY (col1, col2)
    -> )
    ->
    -> PARTITION BY RANGE( col1 ) (
    ->   PARTITION p1 VALUES LESS THAN (11),
    ->   PARTITION p2 VALUES LESS THAN (15)
    -> );
Query OK, 0 rows affected (0.01 sec)

mysql> alter table t1 add unique index  p_col1 (col1);

mysql> CREATE TABLE t2 (
    ->     col1 INT NOT NULL,
    ->     col2 DATE NOT NULL,
    ->     col3 INT NOT NULL,
    ->     col4 INT NOT NULL,
    ->     UNIQUE KEY (col1, col3)
    -> )
    -> PARTITION BY RANGE( col1 + col3 ) (
    ->   PARTITION p1 VALUES LESS THAN (11),
    ->   PARTITION p2 VALUES LESS THAN (15)
    -> );
Query OK, 0 rows affected (0.05 sec)

mysql> alter table t2 add unique index  p_col1 (col1);
```

Comment (tiancaiamao) [reply ddl/db_test.go]: This one I'll focus on alter table.
I'll fix create table in another PR, **it has a bug**, not simply the error message. @zimulala @ciscoxll

File: ddl/ddl_api.go (hunk 2 above)
Comment (tiancaiamao) [reply ddl/ddl_api.go]: Done

File: ddl/db_test.go (hunk 1 above)
Comment (zimulala) [reply ddl/db_test.go]: Do we need to create an issue?

File: ddl/partition.go [hunk 3]
```diff
@@ -483,9 +483,42 @@ func extractConstraintsColumnNames(cons []*ast.Constraint) []map[string]struct{}
...
+		partCols = append(partCols, col.ColName.L)
+	}
+	return partCols, nil
+}
+
 // checkConstraintIncludePartKey checks that the partitioning key is included in the constraint.
-func checkConstraintIncludePartKey(partkeys []string, constraints map[string]struct{}) bool {
-	for _, pk := range partkeys {
+func checkConstraintIncludePartKey(partCols []string, constraints map[string]struct{}) bool {
+	for _, pk := range partCols {
```
Comment (zimulala) [inline ddl/partition.go]: s/pk/pc ?

File: ddl/db_test.go (hunk 1 above)
Comment (tiancaiamao) [reply ddl/db_test.go]: https://github.com/pingcap/tidb/pull/9345

File: ddl/partition.go (hunk 3 above)
Comment (tiancaiamao) [reply ddl/partition.go]: Done

**Comment by tiancaiamao** [comment]
PTAL @zimulala @ciscoxll @jackysp

**Comment by zimulala** [comment]
Please fix CI.

**Pull request merged**